brushfire-core/src/main/scala/com/stripe/brushfire/Tree.scala

```diff
@@ -31,6 +32,107 @@ case class Tree[K, V, T](root: Node[K, V, T]) {
+    val vData = mutable.Map() ++ validationData // Create a mutable Map to allow merging of validation data as a side effect of the pruneLevel method.
+    start match {
+      case leaf: LeafNode[K, V, T] => leaf // Bounce at the bottom and start back up the tree.
+      case split: SplitNode[K, V, T] => {
+        val newNode = SplitNode(children = split.children.map { case (k, p, n) => (k, p, pruneNode(validationData, n, voter, error)) })
+        newNode.children.exists {
+          case (k, v, s: SplitNode[K, V, T]) => true // If we have any splits as children, we can't prune.
+          case _ => false // If all children are leaves, we can prune.
+        } match {
+          case true => newNode
+          case false => pruneLevel(newNode, newNode.children.asInstanceOf[Seq[(K, Predicate[V], LeafNode[K, V, T])]], vData, voter, error)
+        }
+      }
+    }
+  }
+
+  /**
+   * Test conditions and optionally replace parent with a new leaf that combines children.
+   *
+   * Also merges validation data for any combined leaves. This relies on a hack that assumes no leaves have negative
+   * indicies to start out.
+   *
+   * @param parent
+   * @param children
+   * @return
+   */
+  def pruneLevel[P, E](parent: SplitNode[K, V, T], children: Seq[(K, Predicate[V], LeafNode[K, V, T])], validationData: mutable.Map[Int, T], voter: Voter[T, P], error: Error[T, P, E])(implicit weightMonoid: Monoid[T], errorOrdering: Ordering[E]): Node[K, V, T] = {
+    // Get training and validation data and validation error for each leaf.
+    val leafData: Seq[(T, T, E)] = children.map {
+      case (k, p, leaf) =>
+        val validationTarget: T = validationData.getOrElse(leaf.index, weightMonoid.zero)
+        val trainingTarget = leaf.target
+        val leafError = error.create(validationTarget, voter.combine(Some(trainingTarget)))
+        (trainingTarget, validationTarget, leafError)
+    }
+    val (targets: Seq[T], validations: Seq[T], errors: Seq[E]) = leafData.unzip3
+    println("targets:", targets)
+    println("validations:", validations)
+    println("errors:", errors)
+    val targetSum: T = weightMonoid.sum(targets) // Combined training targets to create the training data of the potential combined node.
+    val targetPrediction: P = voter.combine(Some(targetSum))
+    val validationSum: T = weightMonoid.sum(validations) // Combined validation target for combined node.
+    println("target sum:", targetSum)
+    println("target prediction:", targetPrediction)
+    println("validation sum:", validationSum)
+    val errorOfSums: E = error.create(validationSum, targetPrediction) // Error of potential combined node.
+    val sumOfErrors: Option[E] = error.semigroup.sumOption(errors) // Sum of errors of leaves.
+    val doCombine: Boolean = sumOfErrors match {
+      case Some(soe) => errorOrdering.gt(soe, errorOfSums)
```

Comment:
this should be `gte` not `gt` - if the errors are equal, we might as well combine